core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -73,221 +112,818 @@ public static RecordState forId(byte id) {
+                        lastBatch.lastOffset(), lastBatch.nextOffset()));
+                }
+                future.complete(result);
             }
-            nextFetchOffset = next;
+        } finally {
+            lock.writeLock().unlock();
         }
+        return future;
     }
 
+    /**
+     * Acknowledge the fetched records for the share partition. The accepted batches are removed
+     * from the in-flight records once persisted. The next fetch offset is updated to the next offset
+     * that should be fetched from the leader, if required.
+     *
+     * @param memberId The member id of the client that is fetching the record.
+     * @param acknowledgementBatch The acknowledgement batch list for the share partition.
+     *
+     * @return A future which is completed when the records are acknowledged.
+     */
     public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<AcknowledgementBatch> acknowledgementBatch) {
-        log.debug("Acknowledgement batch request for share partition: " + acknowledgementBatch);
-        throw new UnsupportedOperationException("Not implemented yet");
+        log.trace("Acknowledgement batch request for share partition: {}-{}", groupId, topicIdPartition);
+
+        Throwable throwable = null;
+        lock.writeLock().lock();
+        List<InFlightState> updatedStates = new ArrayList<>();
+        try {
+            long localNextFetchOffset = nextFetchOffset;
+            // Avoided using enhanced for loop as need to check if the last batch have offsets
+            // in the range.
+            for (int i = 0; i < acknowledgementBatch.size(); i++) {
+                AcknowledgementBatch batch = acknowledgementBatch.get(i);
+
+                RecordState recordState;
+                try {
+                    recordState = fetchRecordState(batch.acknowledgeType);
+                } catch (IllegalArgumentException e) {
+                    log.debug("Invalid acknowledge type: {} for share partition: {}-{}",
+                        batch.acknowledgeType, groupId, topicIdPartition);
+                    throwable = new InvalidRequestException("Invalid acknowledge type: " + batch.acknowledgeType);
+                    break;
+                }
+
+                // Find the floor batch record for the request batch. The request batch could be
+                // for a subset of the batch i.e. cached batch of offset 10-14 and request batch
+                // of 12-13. Hence, floor entry is fetched to find the sub-map.
+                Map.Entry<Long, InFlightBatch> floorOffset = cachedState.floorEntry(batch.baseOffset);
+                if (floorOffset == null) {
+                    log.debug("Batch record {} not found for share partition: {}-{}", batch, groupId, topicIdPartition);
+                    throwable = new InvalidRequestException("Batch record not found");
```

Comment:
I wonder whether it might be worth having different exception string for these cases so we can differentiate between them during problem determination where tracing is not turned on.

Reply:
Added different messages, we might need to define them better later on.